processor/tailsamplingprocessor/internal/sampling/ottl.go

```diff
@@ -0,0 +1,132 @@
+
+import (
+	"context"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/filter/expr"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/filter/filterottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlspan"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlspanevent"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+type ottlStatementFilter struct {
+	sampleSpanExpr      expr.BoolExpr[ottlspan.TransformContext]
+	sampleSpanEventExpr expr.BoolExpr[ottlspanevent.TransformContext]
+	ErrorMode           ottl.ErrorMode
+	logger              *zap.Logger
+}
+
+var _ PolicyEvaluator = (*ottlStatementFilter)(nil)
+
+// evalFuncChain contains eval functions for:
+// 1. span expression
+// 2. span event expression
+var evalFuncChain = []func(ctx context.Context, osf *ottlStatementFilter, span ptrace.Span, scope pcommon.InstrumentationScope, resource pcommon.Resource) (Decision, error){
+	evalSpan,
+	evalSpanEvent,
+}
+
+// NewOTTLStatementFilter looks at the trace data and returns a corresponding SamplingDecision.
+func NewOTTLStatementFilter(logger *zap.Logger, spanStatements, spanEventStatements []string, errMode ottl.ErrorMode) (PolicyEvaluator, error) {
+	filter := &ottlStatementFilter{
+		ErrorMode: errMode,
+		logger:    logger,
+	}
+	var err error
+	if filter.sampleSpanExpr, err = filterottl.NewBoolExprForSpan(spanStatements, filterottl.StandardSpanFuncs(), ottl.PropagateError, component.TelemetrySettings{Logger: zap.NewNop()}); err != nil {
+		return nil, err
+	}
+	if filter.sampleSpanEventExpr, err = filterottl.NewBoolExprForSpanEvent(spanEventStatements, filterottl.StandardSpanEventFuncs(), ottl.PropagateError, component.TelemetrySettings{Logger: zap.NewNop()}); err != nil {
+		return nil, err
+	}
+
+	return filter, nil
+}
+
+func (osf *ottlStatementFilter) Evaluate(ctx context.Context, _ pcommon.TraceID, trace *TraceData) (Decision, error) {
+	osf.logger.Debug("Evaluating with OTTL statement filter")
+
+	if osf.sampleSpanExpr == nil && osf.sampleSpanEventExpr == nil {
+		return NotSampled, nil
+	}
+
+	trace.Lock()
+	batches := trace.ReceivedBatches
+	trace.Unlock()
+
+	for i := 0; i < batches.ResourceSpans().Len(); i++ {
+		rs := batches.ResourceSpans().At(i)
+		resource := rs.Resource()
+		for j := 0; j < rs.ScopeSpans().Len(); j++ {
+			ss := rs.ScopeSpans().At(j)
+			scope := ss.Scope()
+			for k := 0; k < ss.Spans().Len(); k++ {
+				// Now we reach span level and execute each eval function.
+				// The execution of function chain will break when:
+				// 1. error happened.
+				// 2. "Sampled" decision made.
+				// Otherwise, it will keep executing and finally exit with "NotSampled" decision.
+				for l := range evalFuncChain {
```

Comment:
Since we have access to the skipExpr here via `osf` I think we can remove the need for `evalFuncChain`. Checkout how the filterprocessor handles span and spanevent conditions: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/af0f695997043ac28db2e2662ffd28fc799bb8cb/processor/filterprocessor/traces.go#L95-L115

Reply:
OKie it's done inside the loop now. I feel a little bit uncomfortable when writing down so many for loops (lmao). Is it common when processing span-level stuff? It's identical to the filter processor before but I moved them outside the loop for to make it neater...

Reply:
Yup, loops within loops is a very common pattern in processors when accessing data.  It's also slightly more efficient than your original approach which I believe looped through all the spans once for the Span Conditions and then looped through all the spans again for the Span Event conditions.  The filterprocessor method only loops through the spans 1 time.